src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -182,20 +182,63 @@ impl<Src: OpendalSource> OpendalReader<Src> {
             // note that the buffer contains the newline character
             debug_assert_eq!(n_read, line_buf.len());
 
-            // FIXME(rc): Here we have to use `offset + n_read`, i.e. the offset of the next line,
-            // as the *message offset*, because we check whether a file is finished by comparing the
-            // message offset with the file size in `FsFetchExecutor::into_stream`. However, we must
-            // understand that this message offset is not semantically consistent with the offset of
-            // other source connectors.
-            let msg_offset = (offset + n_read).to_string();
-            batch.push(SourceMessage {
-                key: None,
-                payload: Some(std::mem::take(&mut line_buf).into_bytes()),
-                offset: msg_offset,
-                split_id: split.id(),
-                meta: SourceMeta::Empty,
-            });
-            offset += n_read;
+            // The `offset` variable serves two main purposes:
+            // 1. It is used to check whether a file has been completely read. In the fetch executor,
+            //    when `offset >= split.size`, it indicates that the end of the file has been reached,
+            //    prompting the removal of the corresponding entry from the state table.
+            //
+            // 2. After recovery occurs, `offset` prevents the reading process from starting
+            //    over from the beginning of the file. Instead, it allows the reading to continue
+            //    directly from the last successfully read `offset`.
+            //
+            // Special handling is required for files compressed with gzip:
+            // - Gzip files must always be read from the beginning; they cannot be read from
```

Comment:
So we cannot achieve exactly once if gzip enabled?

Reply:
I think so.

Reply:
Good point. I think exactly once is important for source. Some ideas to make reading gzip files exactly once:
1. Yield one large chunk per gzip file containing all contents after decompression. This ensures that no barrier can be placed in the middle of the file but will increase memory usage.
2. Pause fetcher upstream while reading a gzip file. This also ensures that no barrier can be placed in the middle.
3. Use offset and size after decompression to track gzip file fetch progress, like what we already did right now. Opendal reader always starts from the beginning and skip contents below the recorded offset.

Personally perfer 3.

Reply:
I also prefer 3 because it has the least impact.

Reply:
Resolved, had some offline discussion with @hzxa21 .
Please see the description of the PR for specific details. I modified some of the lister and fetcher logic. cc @tabVersion This part was designed by you, for awareness.